src/servers/src/http/authorize.rs

```diff
@@ -0,0 +1,282 @@
+// Copyright 2022 Greptime Team
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::marker::PhantomData;
+
+use axum::http::{self, Request, StatusCode};
+use axum::response::Response;
+use common_telemetry::error;
+use futures::future::BoxFuture;
+use http_body::Body;
+use snafu::{OptionExt, ResultExt};
+use tower_http::auth::AsyncAuthorizeRequest;
+
+use crate::auth::{Identity, UserInfo, UserProviderRef};
+use crate::error::{self, Result};
+
+pub struct HttpAuth<RespBody> {
+    user_provider: Option<UserProviderRef>,
+    _ty: PhantomData<RespBody>,
+}
+
+impl<RespBody> HttpAuth<RespBody> {
+    pub fn new(user_provider: Option<UserProviderRef>) -> Self {
+        Self {
+            user_provider,
+            _ty: PhantomData,
+        }
+    }
+}
+
+impl<RespBody> Clone for HttpAuth<RespBody> {
```

Comment:
Why hand impl `Clone` instead of derived it?

Reply:
if use `#[derive(Clone)]`, the compiler will report an error.

Reply:
Need to manually impl `Clone` for `HttpAuth<RespBody>`, because #[derive(Clone)] cannot be sure that `RespBody` will impl `Clone`. When manually impl `Clone`,  will tell the compiler that it doesn't matter.

Here is [playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=2a9b59d2b59edf02ea4469a7c100985e).